### ISSUE TITLE
primary_param_names is wrong in both directions on a .net model: it omits literal-valued constants and lists every function name

### ISSUE THREAD

**Issue by wshlavacek** [lanl/bngsim#227]
Found while fixing #181. Not the same defect — #181 was the codegen `.net`
parser reading the kind-annotation *comment*; this is the loader's own
classification, and it reaches the Python API rather than the emitted C.

## Summary

`Model.primary_param_names` is the accessor whose own docstring says

> Use this when you want to expose the model to an external optimizer or
> sampler that should treat each parameter as an independent variable

and it is wrong in both directions on a `.net` model. `bngsim.jax.differentiable_solve`
takes it as the default differentiation set (`flat=False`), so both errors land
in a gradient.

`benchmarks/models/net/ode/SIR.net` — 23 lines, BNG2.pl 2.9.3 — shows both at once:

```
begin parameters
    1 S0     2e7   # Constant
    2 I0     1     # Constant
    3 beta   1/S0  # ConstantExpression
    4 gamma  1/7   # Constant
end parameters
begin functions
    1 betaI() beta*I
end functions
```

```python
m = bngsim.Model.load("benchmarks/models/net/ode/SIR.net")
m.param_names          # ['S0', 'I0', 'beta', 'gamma', 'betaI']
m.param_is_expression  # [False, False, True, True, False]
m.primary_param_names  # ['S0', 'I0', 'betaI']
```

`gamma`, the recovery rate, is gone. `betaI`, the *function*, is there.

```python
p0 = jnp.array([m.get_param(n) for n in m.primary_param_names])
jax.grad(loss)(p0)
# {'S0': 2.18e8, 'I0': -2.07e12, 'betaI': 0.0}
```

## Direction 1 — a literal-valued constant is omitted

`net_file_loader.cpp:160-171` decides `is_expression` by whether `std::stod`
consumes the whole value token. `1/7` does not, so `gamma` is "derived" and
`primary_param_names` drops it.

But `gamma` has nothing behind it to be derived *from*. It is a working knob by
every other measure:

* `set_param("gamma", 0.5)` moves the trajectory — `[1.0e5, 4.8e6, 1.5e7]` →
  `[1.6e7, 1.9e6, 2.5e6]` at t=30;
* `Simulator(..., sensitivity_params=["gamma"])` is accepted and returns a live
  column, `max|S| = 8.7e7`. **Both codegen paths already handle it** — I measured
  the `.net`-text path and the model path on `Repressilator.net`'s
  `c6 = ln(2)/120` and got 3.825e6 from each. Nothing downstream is broken; the
  parameter is simply not on the list you are told to hand an optimizer.

Incidence over the 1,817 `.net` files in the tree: **626 parameter lines across
191 models**, of which **535 are referenced somewhere** (240 by another
parameter's expression, 138 by a function, 50 by a rate law, 50 by both a rate
law and an expression, 26 by an initial condition, the rest by combinations).
The shape is always the same — a constant written as arithmetic:

| model | parameter |
|---|---|
| `SIR.net` | `gamma = 1/7` |
| `Repressilator.net` | `c6 = ln(2)/120`, `c7 = ln(2)/600` |
| `RasRaf_stdBNGL_v3.net` | `Temp = 37+273.15`, `pi = 2*asin(1)` |
| `rec_dim_comp.net` | `NaV = 6.022e23*1e-15` |
| `Scaff_22_ground.net` | `Ctot = 5*10e5` |

**BNG2.pl already draws this line correctly** and annotates every one of them
`# Constant`, reserving `# ConstantExpression` for a value that references
another parameter — I checked all 41,433 annotated parameter lines in the tree
and found no counterexample. #181 (PR #226) adopted that rule for the codegen
`.net` parser: *derived iff the expression references another declared
parameter*. The loader still uses the `stod` rule, so the two now disagree on
these 626 lines — harmlessly, since they answer different questions, but it is
the same distinction and only one place gets it right.

**SBML is not affected**: 0 of 248 corpus models carry a parameter that is
`is_expression=True` with a referenceless expression. The probe is not vacuous —
75 of those 248 do carry derived parameters.

## Direction 2 — every function name is listed as a knob

`model_builder.cpp:1246-1263` gives each function a synthetic backing parameter
slot so the parameter vector can hold its evaluated value:

```cpp
Parameter synth;
synth.name = impl.functions[fi].name;
synth.value = 0.0;
synth.expression = "";
synth.is_expression = false;   // <- so primary_param_names lists it
```

`primary_param_names` filters on `not is_expression and not is_internal`, and
this slot sets neither. So:

* `m.get_param("betaI")` is `0.0`, not the function's value;
* `m.set_param("betaI", 3.0)` is **accepted**, `get_param` returns `3.0`, and the
  trajectory does not move;
* `jax.grad` spends a coordinate on it and returns exactly `0.0`, forever.

**327 of 327** function-carrying `.net` models sampled leak every function name
this way — it is universal, not an edge case. Examples: `Kholodenko_2000.net`
(`v1`, `v2`, `v10`, …), `CaOscillate_functional.net` (`sat3`, `sat5`, `sat11`),
`RAFi_ground.net` (`Activity`, `Ybar`).

#170 already introduced `is_internal` for exactly this category — a synthesized
slot that is not a knob of the model (`_V0_<comp>`) — and `primary_param_names`
already excludes on it. The function slot looks like the same case with the flag
left unset, but I have not checked what else reads `is_internal`, so that is a
suggestion and not a diagnosis.

## Why it matters

A silently-absent parameter and a permanently-zero coordinate both survive every
check a fitting frontend can make locally: the objective looks right, the
gradient check passes on the coordinates that *are* present, and only the search
behaves oddly. On `SIR.net` the default `jax.grad` path fits the two initial
conditions and one dead coordinate, and holds the recovery rate fixed — with no
warning.

## Not proposed here

Both fixes change `primary_param_names`, which is public and is what
`differentiable_solve`'s `flat=False` vector is ordered by, so a fix changes
gradient vector *length and order* for the 191 + 327 affected models. Whether
that is a MAJOR bump, and whether direction 1 should move the loader's
`is_expression` or only the `primary_param_names` filter, are calls for the
maintainer.

Measurements above are reproducible on `main` at `22def5d`; the corpus figures
use the materialized `parity_checks/rr_parity/models` tree.
